text/0000-associated-type-lifetime-elision.md

```diff
@@ -0,0 +1,348 @@
+is using lifetime elision, so elided lifetime parameters (other than with `&`
+or `&mut`) must be marked with `'_`, such as `type Output = Ref<'_, T>;`.
+Omitting a lifetime parameter, such as `type Output = Ref<T>;`, is an error.
+(These are the same rules as those checked by the `elided_lifetimes_in_paths`
+lint, which is part of the `rust_2018_idioms` group.)
+
+This functionality will be available in the 2015 and following editions.
+
+Examples:
+
+```rust
+impl Substr<u32> for &str { type Output = &str; }                // elided
+fn substr(s: &str, until: u32) -> &str;                          // fn analog
+impl<'a> Substr<u32> for &'a str { type Output = &'a str; }      // expanded
+
+impl Finder<&str> for u32 { type Output = &str; }                // elided
+fn finder(x: u32, y: &str) -> &str;                              // fn analog
+impl<'a> Finder<&'a str> for u32 { type Output = &'a str; }      // expanded
+
+impl GetStr for u32 { type Output = &str; }                      // ILLEGAL
+fn get_str(x: u32) -> &str;                                      // fn analog
+// No input position
+
+impl Frob<&str> for &str { type Output = &str; }                 // ILLEGAL
+fn frob(x: &str, y: &str) -> &str;                               // fn analog
+// `Self` isn't special
+
+impl GetMut for &mut Foo { type Output = &mut Bar; }             // elided
+fn get_mut(x: &mut Foo) -> &mut Bar;                             // fn analog
+impl<'a> GetMut for &'a mut Foo { type Output = &'a mut Bar; }   // expanded
+
+impl New for &mut [u8] { type Output = BufWriter<'_>; }          // elided
+fn new(buf: &mut [u8]) -> BufWriter<'_>;                         // fn analog
+impl<'a> New for &'a mut [u8] { type Output = BufWriter<'a>; }   // expanded
+
+impl New for &mut [u8] { type Output = BufWriter; }              // ILLEGAL
+// Hidden lifetime parameter
+
+impl Two for &str { type A = &str; type B = &str; }              // elided
+fn two(x: &str) -> (&str, &str);                                 // fn analog
+impl<'a> Two for &'a str { type A = &'a str; type B = &'a str; } // expanded
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+As this is an extension of elision, it shares the same drawbacks: there's no
+lifetime name to mention in error messages, it delays a rigorous understanding
+of lifetimes, the rules need to be understood to comprehend a signature, etc.
+The trade-off has proven worth-while for `fn`, however, and re-using the same
+rules here should keep the learning cost low.
+
+This does make associated types somewhat special.  Notably, the input positions
+from the impl header don't apply to `fn` return type outputs, so something like
+the following doesn't work, though one might think it should:
+
+```rust
+impl Foo<&str> for u32 {
+    fn bar(self) -> &str; // ERROR: missing lifetime specifier
+    // help: this function's return type contains a borrowed value,
+    // but there is no value for it to be borrowed from
+}
```

Comment:
Similarly, the `IntoIterator` pattern also falls down here if you attempt to mention the type by-name in return position, rather than using the `Self::` form:

```rust
impl<T> IntoIterator for &SkipList<T> {
    type Item = &T;
    type IntoIter = Iter<'_, T>;

    // works:
    fn into_iter(self) -> Self::IntoIter { self.iter() }

    // doesn't work:
    fn into_iter(self) -> Iter<'_, T> { self.iter() }
}
```

This doesn't seem that bad to me, especially since we already tend to push users towards writing `Self::...` rather than naming the type directly for other reasons (e.g. DRY), but it does seem a bit unfortunate. It's another point in the "type aliases aren't syntactic substitutions" field.

Reply:
There's an input-position equivalent of this problem too:
```rust
impl<'a> From<&'a str> for String {
    fn from(s: &'a str) -> String {
        s.to_owned()
    }
}
```
That's an input-position lifetime that would be great to elide, since it's never used in output position and doesn't tie any input lifetimes together, but it can't be because `fn from(s: &str) -> String` is something different.

This makes me think of the same "different syntax for lifetimes at different levels" discussion that was had about in-band lifetimes: With strawman syntax of two apostrophes for an up-level lifetime, this could work:
```rust
impl From<&str> for String {
    fn from(s: &''_ str) -> String {
        s.to_owned()
    }
}
```

But I'd rather not include something like that in this RFC.

Reply:
While I agree that `&''` should not be in this RFC (and I'm not sure I'm a fan...) I think it would be good to include it in a section on "possible future work" or something to record the idea in a more visible location than GitHub comments.

I think Taylor's notes should also be discussed somewhere in the RFC. :)